### ISSUE TITLE
WindowsFormsMessagePump and WpfMessagePump attributes

### ISSUE THREAD

**Issue by jnm2** [nunit/nunit#3088]
tl;dr We haven't fully addressed Windows Forms and WPF message pumping. Even though we're planning to make this user-extensible, we should still ship these attributes out of the box.

### Problem

Windows Forms needs to run on an STA thread. If you add `[Apartment(System.Threading.ApartmentState.STA)]` to your test, we install a [single-threaded synchronization context](https://github.com/nunit/nunit/blob/208adcefd49cedfa1b7500800348ac7ce300f526/src/NUnitFramework/framework/Internal/SingleThreadedSynchronizationContext.cs#L30) before invoking async methods so that [awaits from an STA thread continue on the same thread](https://github.com/nunit/nunit/issues/1200). This is pretty important and NUnit is doing the right thing now.

Here's the problem. A side effect of having a derived synchronization context is that instantiating a `System.Windows.Forms.Control` or derived type will no longer auto-install a WindowsFormsSynchronizationContext. ([Source](https://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/WinForms/WindowsFormsSynchronizationContext.cs,139))—oddly enough, I [copied their logic exactly](https://github.com/nunit/nunit/blob/208adcefd49cedfa1b7500800348ac7ce300f526/src/NUnitFramework/framework/Internal/AsyncToSyncAdapter.cs#L119) without knowing it.) And presence of a WindowsFormsSynchronizationContext is what NUnit looks for to know whether to pump messages during awaits using `System.Windows.Forms.Application`. In order to gain this back, users have to do `SynchronizationContext.SetSynchronizationContext(new WindowsFormsSynchronizationContext());` before the await that pumps messages.

And with or without WindowsFormsSynchronizationContext autoinstalling itself, the only way to get  NUnit to pump messages while awaiting the test method itself is to set the synchronization context in a `SetUp` method. (And if the SetUp method itself is async, there's no way to pump messages while NUnit awaits it.)

### Proposed solution

Tests that use Windows Forms and WPF should replace `[Apartment(ApartmentState.STA)]` (which will now be implied) with `[WindowsFormsMessagePump]` or `[WpfMessagePump]` which is shorter and more idiomatic.

NUnit will see these and install the appropriate synchronization context before each test that is marked this way. SetUp, TearDown, and the test method will all have messages pumped while being awaited.
Ideally I think they should be applicable to fixtures and should cause messages to be pumped while awaiting OneTimeSetUp and OneTimeTearDown.

The implementation may be a good way to kick the tires of the user-extensible test running mechanism we want to build.

**Comment by JohanLarsson** [comment]
Changing the attribute is a code fix opportunity. Not a huge win over find & replace though.

**Comment by jnm2** [comment]
There is probably a better name for these attributes around here somewhere, too. The test method itself isn't a message pump. `[WindowsFormsTest]` might be better.

**Comment by jnm2** [comment]
Here is a [working proof-of-concept](https://gist.github.com/jnm2/d10e71a3e10a000692b17afc641b042a) that relies on NUnit 3.11’s special recognition of WindowsFormsSynchronizationContext.

**Comment by ChrisMaddock** [comment]
Sounds ok to me. 👍 

I'm understanding correctly, that all functionality required here is in the new attributes? So a user who wished to support a new GUI kit/synchronization context for whatever reason down the line, could just create their own custom attribute? 🙂
